### ISSUE TITLE
Tilemap world_to_map() does not get the correct tile in combination with Raycast2D

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot#35344]
**Godot version:** 3.1 stable, 3.2 rc1

**OS/device including version:** Win7

**Issue description:**
When casting a ray onto a tilemap to detect a tile, I would expect for world_to_map() to return the Tile behind the ray collision point, not on the right side or below it. 
![tileset_mining](https://user-images.githubusercontent.com/47016402/72697166-39632a00-3b3f-11ea-866b-ff9794c25910.gif)

**Steps to reproduce:**
```
extends RayCast2D

func _physics_process(delta):
	if Input.is_action_just_pressed("Mine_Left") and is_colliding():
		var target = get_collider()
		if target.is_class("TileMap"):
			var point = get_collision_point()
			var map_point = target.world_to_map(point)
			target.set_cellv(map_point, 0, false, true)
```

**Minimal reproduction project:**
[Tileset_Mining.zip](https://github.com/godotengine/godot/files/4084345/Tileset_Mining.zip)

**Comment by groud** [comment]
The behavior is kind of expected then. Instead of using the collision point, which is at the border of the tile, you should probably use the tip of your raycast instead.

**Comment by Xrayez** [comment]
I guess you could also do:
```gdscript
var safe_margin = 1.0 # increase this if doesn't work
var point = get_collision_point() - get_collision_normal() * safe_margin
```
That way you can kind of prevent these edge-cases where floating point math is not your friend.

Might be also related to something similar issue as in #26615.

**Comment by golddotasksquestions** [comment]
@Xrayez Thank you so much, this works great!
If that's the expected way to solve this problem, it probably should be documented somewhere.

**Comment by Xrayez** [comment]
@golddotasksquestions maybe, yet this specific problem doesn't seem to pertain to tilemap or raycast in isolation so not sure if this can be properly documented, at least in the class reference. I find it difficult to come up with proper wording to describe this, so it does feel like a workaround currently.

If you re-run the raycast pointing from the origin to the collision point again *exactly*, `is_colliding` should return `true`. I know that because I do quite similar technique for a use case where I want to *prevent* continuous raycast detection, the only difference is that `-` becomes `+`:

```gdscript
var point = get_collision_point() + get_collision_normal()
```

It means that the returned collision point *cannot* be blamed for inaccurate results. So yeah it does smell like `world_to_map()` is not precise enough.

I've also looked at the source code for this method now, and stumbled upon this issue: #23250. So the proposed fix is indeed a hack currently. Your use case was described by @Zylann at https://github.com/godotengine/godot/issues/23250#issuecomment-435162860.

**Comment by golddotasksquestions** [comment]
> It means that the returned collision point cannot be blamed for inaccurate results.

Yes, exactly. In the Minimal project I moved the ColorRect to the collision point position to verify this. It's hundred percent accurate.

I've also skimmed over #23250 before writing this, but I did not understood the "hack" and assumed because it was closed 23250 was about something else.

**Comment by nikhilCad** [comment]
@Xrayez thanks for the fix above, you just saved my life!

**Comment by KoBeWi** [comment]
Looks still valid in 4.4. Roughly updated MRP:

[Tileset_Mining.zip](https://github.com/user-attachments/files/18825478/Tileset_Mining.zip)

It's marked as discussion. Are there plans to do something about it?

**Comment by groud** [comment]
> It's marked as discussion. Are there plans to do something about it?

Not really. As I mentioned, the returned collision point is the exact point where the raycast and the tile's shape collide, which is on the border of the tile. And since it's on the edge of the tile, there's some precision issue which make retrieving the tile unreliable.

If we want to solve the issue (which will likely be even more problematic once collision shapes merging gets implemented), I think we might need to implement a custom helper. Likely taking the collision data as input and returning the collided tile(s).
